packages/angular_devkit/core/src/jobs/architecture.md

```diff
@@ -0,0 +1,253 @@
+functionality on top of RxJS' `Observable`s.
+
+# Background
+An `Observable` (at a higher level) is a function that receives a `Subscriber`, and outputs
+multiple values, and finishes once it calls the `Subscriber.prototype.complete()` method (in
+JavaScript):
+
+```javascript
+const output1To10EverySecond = function (subscriber) {
+  let t = 0;
+  const i = setInterval(() => {
+    t++;
+    subscriber.next(t);
+    if (t === 10) {
+      subscriber.complete(t);
+    }
+  }, 1000);
+  return () => clearInterval(i);
+};
+
+const stream$ = new Observable(output1To10EverySecond);
+// Start the function, and output 1 to 100, once per line.
+stream$.subscribe(x => console.log(x));
+```
+
+This, of course, can be typed in TypeScript, but those types are not enforced at runtime.
+
+# Glossary
+- `job handler`. The function that implements the job's logic.
+- `raw input`. The input observable sending messages to the job. These messages are of type
+    `JobInput`.
+- `raw output`. The output observer returned from the `job handler`. Messages on this observable
+    are of type `JobEvent`.
+
+# Description
+
+A `JobHandler`, similar to observables, is a function that receives an argument and a context, and
+returns an `Observable` of events, which can include outputs that are typed at runtime (using a
+Json Schema):
+
+```javascript
+const output1ToXEverySecond = function (x, context) {
+  return new Observable(subscriber => {
+    let t = 0;
+
+    // Notify our users that the actual work is started.
+    subscriber.next({ kind: JobEventKind.Start });
+    const i = setInterval(() => {
+      t++;
+      subscriber.next({ kind: JobEventKind.Output, value: t });
+      if (t === x) {
+        subscriber.next({ kind: JobEventKind.End });
+        subscriber.complete();
+      }
+    }, 1000);
+
+    return () => {
+      clearInterval(i);
+    };
+  })
+};
+
+// For now, jobs can not be called without a registry and scheduler.
+const registry = new SimpleJobRegistry();
+registry.register('output-from-1-to-x', output1ToXEverySecond, {
+  argument: { type: 'number' },
+  output: { type: 'number' },
+});
+const scheduler = new SimpleScheduler(registry);
+
+// Need to keep the same name that the registry would understand.
+// Count from 1 to 10.
+const job = scheduler.schedule('output-from-1-to-x', 10);
+
+// A Job<> instance has more members, but we only want the output values here.
+job.output.subscribe(x => console.log(x));
+```
+
+This seems like a lot of boilerplate in comparison, but there are a few advantages;
+
+1. lifecycle. Jobs can tell when they start doing work and when work is done.
+1. everything is typed, even at runtime.
+1. the context also contains an input Observable that receives typed input events, including
+    input values, and stop requests.
+1. jobs can also schedule other jobs and wait for them, even if they don't know if a job is
+    implemented in the system.
+
+## Diagram
```

Comment:
Can you also explain in this section the relationship between `Input`, `RawInput`, `Output`, and `RawOutput`?

I've read this section, the 3 relevant sections in the README ("Argument, Input, Output and Channels", "Job<OutputType> Object", and "Communicating With Jobs") and I still don't get those.

These are the things I don't understand:
- Why are there two? (e.g. both `input` and `rawInput`)
- When I have a `Job<OutputType>` object, and I want to send messages to the job handler, which one should I use? (and same for receiving messages)
- What do I use the other one for?
- In a `JobHandlerContext<I, O>` object, `rawInput` is the complement of `rawInput` from a  `Job<OutputType>`. What's the complement of `input`?
- Are all these validated, or only some?

And example of using all of these would be useful.

Reply:
> Why are there two? (e.g. both input and rawInput)

```
input = rawInput.pipe(
    filter(event => event.kind === JobInputKind.Input),
    map(event => event.value),
);
```

This is to reduce boilerplate.

> When I have a Job<OutputType> object, and I want to send messages to the job handler, which one should I use? (and same for receiving messages)
Are you sending raw input events or just input values? e.g. are you sending `ping()`s? 

It's a leaky abstraction, by design. `rawInput/rawOutput` are the leaks. They're needed because in some cases you just want to forward a job's I/O untransformed. Or add to them. The other alternative was to kill `input` and `output` which means that everytime you want to output a value, instead of doing `return Promise.resolve(value)` you'd have to do `rawOutput.next({ kind: JobEventKind.Output, description: context.description, value })` which is a lot of boilerplate.

> What do I use the other one for?
See above.

> In a JobHandlerContext<I, O> object, rawInput is the complement of rawInput from a Job<OutputType>. What's the complement of input?
`createJobHandler` gets the `output` from the return value of the handler. Without `createJobHandler` it would be a lot of boilerplate (as seen in examples).

> Are all these validated, or only some?
At this point everything is validated. Output and Channels will error out if there's an error, `input` will just filter out values that are invalid.

Reply:
This is a very confusing abstraction... To say "`input` are the mapped values of `rawInput` filtered by `JobInputKind.Input`" has the word "input" a lot of times. One of these probably should be called something else.

I think that input is a overloaded term because it means these two different things:
- `input` is the way you send values to a job. `job.input.next(1);` sends an event of kind `JobInputKind.Input`.
- `rawInput` contains events of many kinds, one of which is `JobInputKind.Input`. When you are composing jobs you want direct access to all of these events.

So `rawInput` sounds like an event bus that is prefiltered by events of the type `JobInput`, and `rawOutput` is also an event bus that only has events of the type `JobEvent`. It's also confusing that `JobEvent` is not `JobOutput`. I don't have an opinion on whether two event buses are needed though.

If I had to pick a super descriptive names I would rename:
- `JobInput` to `InboundEvent`
- `rawInput` to `InboundEventBus`
- `JobEvent` to `OutbountEvent`
- `rawOutput` to `OutbountEventBus`

Then I could say "`input` are the mapped values of `InboundEventBus` filtered by `InboundEventKind.Input`", which I think is clearer.

I'm not requesting you use these names though. But I am highlighting that the names `input` and `output` are overloaded and do a poor job of communicating their meaning and function.

Reply:
Let me think a bit more about this.

Reply:
What do you think of InboundMessage and OutboubdMessage instead of event?

Reply:
SGTM

Reply:
Done.